src/solvers/twostep_time_solver.C

```diff
@@ -257,4 +287,103 @@ void TwostepTimeSolver::solve()
     }
 }
 
+std::pair<unsigned int, Real> TwostepTimeSolver::adjoint_solve (const QoISet & qoi_indices)
+{
+  // The adjoint timestepping mirrors the scheme used for the forward problem
+  // So the deltat, once set by solution history, will not be changed
+  Real old_time = _system.time;
+
+  // Take the first adjoint 'half timestep'
+  core_time_solver->adjoint_solve(qoi_indices);
+
+  // Adjoint advance the timestep
+  core_time_solver->adjoint_advance_timestep();
+
+  // The second half timestep
+  std::pair<unsigned int, Real> full_adjoint_output = core_time_solver->adjoint_solve(qoi_indices);
+
+  // Tell our timesolver the combined deltat before adjoint advance timestep updates it
+  this->completedtimestep_deltat = 2.0*_system.deltat;
+
+  // Adjoint advance again so solution history saves the adjoint
+  core_time_solver->adjoint_advance_timestep();
+
+  // Reset the system.time
+  _system.time = old_time;
+
+  return full_adjoint_output;
+}
+
+void TwostepTimeSolver::integrate_adjoint_sensitivity(const QoISet & qois, const ParameterVector & parameter_vector, SensitivityData & sensitivities)
+{
+  // We are using the midpoint rule to integrate each timestep
```

Comment:
We have core_time_solver->integrate_adjoint_sensitivity() available now, right?  So let's use that instead, and get the right answer even if we're using higher-order time stepping?

Reply:
Yeah, will need to be careful about the detat's, but you are right.